src/library/scala/math/BigInt.scala

```diff
@@ -221,84 +252,83 @@ final class BigInt private (private val _bigInteger: BigInteger, private val _lo
+      if (that.longEncoding) false else this._bigInteger == that._bigInteger
     }
-    )
 
   /** Compares this BigInt with the specified BigInt
    */
   def compare (that: BigInt): Int =
-    if (this.isValidLong) {
-      if (that.isValidLong) java.lang.Long.compare(this._long, that._long) else -that._bigInteger.signum()
+    if (this.longEncoding) {
+      if (that.longEncoding) java.lang.Long.compare(this._long, that._long) else -that._bigInteger.signum()
     } else {
-      if (that.isValidLong) this.signum else this._bigInteger.compareTo(that._bigInteger)
+      if (that.longEncoding) this.signum else this._bigInteger.compareTo(that._bigInteger)
     }
 
   /** Addition of BigInts
    */
   def +  (that: BigInt): BigInt = {
-    if (this.isValidLong && that.isValidLong) { // fast path
+    if (this.longEncoding && that.longEncoding) { // fast path
       val x = this._long
       val y = that._long
       val z = x + y
-      if ((~(x ^ y) & (x ^ z)) >= 0L) return BigInt(z)
+      if ((~(x ^ y) & (x ^ z)) >= 0L) new BigInt(z)
     }
     new BigInt(this.bigInteger.add(that.bigInteger))
   }
 
   /** Subtraction of BigInts
    */
   def -  (that: BigInt): BigInt = {
-    if (this.isValidLong && that.isValidLong) { // fast path
+    if (this.longEncoding && that.longEncoding) { // fast path
       val x = this._long
       val y = that._long
       val z = x - y
-      if (((x ^ y) & (x ^ z)) >= 0L) return BigInt(z)
+      if (((x ^ y) & (x ^ z)) >= 0L) new BigInt(z)
     }
     new BigInt(this.bigInteger.subtract(that.bigInteger))
   }
 
   /** Multiplication of BigInts
    */
   def *  (that: BigInt): BigInt = {
-    if (this.isValidLong && that.isValidLong) { // fast path
+    if (this.longEncoding && that.longEncoding) { // fast path
       val x = this._long
       val y = that._long
       val z = x * y
-      if (x == 0 || (y == z / x && !(x == -1 && y == Long.MinValue))) return BigInt(z)
+      if (x == 0 || (y == z / x && !(x == -1 && y == Long.MinValue))) new BigInt(z)
     }
     new BigInt(this.bigInteger.multiply(that.bigInteger))
   }
 
   /** Division of BigInts
    */
   def /  (that: BigInt): BigInt = {
-    if (this.isValidLong && that.isValidLong) { // fast path
-      if (this._long != Long.MinValue || that._long != -1) return BigInt(this._long / that._long)
+    if (this.longEncoding && that.longEncoding) { // fast path
+      if (this._long != Long.MinValue || that._long != -1) new BigInt(this._long / that._long)
     }
     new BigInt(this.bigInteger.divide(that.bigInteger))
   }
 
   /** Remainder of BigInts
    */
   def %  (that: BigInt): BigInt = {
-    if (this.isValidLong && that.isValidLong) { // fast path
-      if (this._long != Long.MinValue || that._long != -1) return BigInt(this._long % that._long)
+    if (this.longEncoding && that.longEncoding) { // fast path
+      if (this._long != Long.MinValue || that._long != -1) new BigInt(this._long % that._long)
```

Comment:
I'm not a fan of these `if` with fall through... pretty dangerous... It seems this is wrong now and the fast path is never hit.

Reply:
Good catch! I corrected those locally, will review the code in the morning before pushing the corrections.